Runtime/Scripts/Timeline/Internal/AnimationTrack.cs

```diff
@@ -2,69 +2,129 @@
+        public void SampleIfChanged(float time) => recordSampleIfChanged(time, sampler.sample(animationData));
+
+        private readonly ProfilerMarker recordSampleIfChangedMarker;
+        private readonly ProfilerMarker unityObjectCheck = new ProfilerMarker("unityObjectCheck");
+        private readonly ProfilerMarker lastSampleCheck = new ProfilerMarker("lastSampleCheck");
+        private readonly ProfilerMarker lastSampleCheckEquality = new ProfilerMarker("lastSampleCheckEquality");
+        private readonly ProfilerMarker removeLastSample = new ProfilerMarker("removeLastSample");
+        private readonly ProfilerMarker insertData = new ProfilerMarker("insert Data");
         
-        protected void recordSampleIfChanged(double time, TData? value) {
-            if (value == null || (value is Object o && !o)) return;
+        protected void recordSampleIfChanged(float time, TData? value) {
+            using var _ = recordSampleIfChangedMarker.Auto();
+            {
+                using var __ = unityObjectCheck.Auto();
+                if (value == null || (value is Object o && !o)) return;
+            }
+            // Additional check to make sure we ignore duplicate timestamps.
+            // If you think, "This doesn't seem necessary, lets remove this": Removing this is _a bad idea_, do not do it.
+            // The resulting animations may otherwise contain the weirdest and seemingly random bugs you can imagine
+            // since duplicate timestamps cause the below "replace consecutive equal samples"-logic to trip over
+            // its own feet and die a horrible death falling down a cliff.
+            // This will cause the duplicated sample value to be recorded, but then removed again entirely and the resulting animation is all kinds of broken:
+            // Consider the following example:
+            // Event 1: Recording sample at time t1 with value v1, samples = [(t1, v1)], lastSample = (t1, v1), secondToLastSample = null
+            // Event 2: Recording sample at time t2 with value v2, samples = [(t1, v1), (t2,v2)], lastSample = (t2, v2), secondToLastSample = (t1, v1)
+            // Event 3: Recording sample at time t2 with value v2, samples = [(t1, v1), (t2,v2)], lastSample = (t2, v2), secondToLastSample = (t2, v2)
+            //          the logic is now in an unexpected state, where it thinks the (t2,v2) sample is not necessary
+            // Event 4: Recording sample at time t3 with value v2, samples = [(t1, v1), (t3,v2)], lastSample = (t3, v2), secondToLastSample = (t2, v2).
+            // If you are wondering, why this can even happen since duplicate timestamps are already checked for in the GLTFRecorder:
+            // Yeah i am not sure why either, but it definitely happens!
+            // Finding this bug cost me 3 days of my life, please don't remove this check.
+            if (time <= LastTime) {
```

Comment:
Good explanation, and horrible that this is even necessary 😭 (you needn't change anything, though)